psh/pshapp/pshapp.c

```diff
@@ -815,30 +823,260 @@ static int psh_parsecmd(char *line, int *argc, char ***argv)
 {
 	char *cmd, *arg, **rargv;
 
-	if ((cmd = strtok(line, "\t ")) == NULL)
+	cmd = strtok(line, "\t ");
+	if (cmd == NULL) {
 		return -EINVAL;
-
-	if ((*argv = malloc(2 * sizeof(char *))) == NULL)
+	}
+	*argv = malloc(2 * sizeof(char *));
+	if (*argv == NULL) {
 		return -ENOMEM;
+	}
 
 	*argc = 0;
 	(*argv)[(*argc)++] = cmd;
 
-	while ((arg = strtok(NULL, "\t ")) != NULL) {
-		if ((rargv = realloc(*argv, (*argc + 2) * sizeof(char *))) == NULL) {
+	for (;;) {
+		arg = strtok(NULL, "\t ");
+		if (arg == NULL) {
+			break;
+		}
+
+		rargv = realloc(*argv, (*argc + 2) * sizeof(char *));
+		if (rargv == NULL) {
 			free(*argv);
 			return -ENOMEM;
 		}
 
 		*argv = rargv;
 		(*argv)[(*argc)++] = arg;
 	}
+
 	(*argv)[*argc] = NULL;
 
 	return 0;
 }
 
 
+static void psh_redirCleanup(struct psh_redir *redir)
+{
+	for (int i = 0; i < PSH_REDIRSZ; ++i) {
+		if (redir->red[i] >= 0) {
+			close(redir->red[i]);
+			redir->red[i] = -1;
+		}
+
+		if (redir->save[i] >= 0) {
+			close(redir->save[i]);
+			redir->save[i] = -1;
+		}
+	}
+}
+
+
+#define PSH_MARKER_IN  (1 << 0)
+#define PSH_MARKER_OUT (1 << 1)
+#define PSH_MARKER_ERR (1 << 2)
+
+
+static int psh_parseRedirections(int *argc, char ***argv, struct psh_redir *redir)
+{
+	int fd, flags, eat, marker;
+	char *path;
```

Comment:
scope of those variables can be limited

Reply:
Only limiting `fd` made sense to me - rest would be just moved under `while`